dfir_rs/src/scheduled/graph.rs

```diff
@@ -274,52 +277,81 @@ impl<'a> Dfir<'a> {
                 // This must be true for the subgraph to be enqueued.
                 assert!(sg_data.is_scheduled.take());
-                tracing::info!(
-                    sg_id = sg_id.to_string(),
-                    sg_name = &*sg_data.name,
-                    sg_depth = sg_data.loop_depth,
-                    "Running subgraph."
-                );
 
                 match sg_data.loop_depth.cmp(&self.context.loop_nonce_stack.len()) {
                     Ordering::Greater => {
                         // We have entered a loop.
                         self.context.loop_nonce += 1;
                         self.context.loop_nonce_stack.push(self.context.loop_nonce);
-                        tracing::warn!(loop_nonce = self.context.loop_nonce, "Entered loop.");
+                        tracing::trace!(loop_nonce = self.context.loop_nonce, "Entered loop.");
                     }
                     Ordering::Less => {
                         // We have exited a loop.
                         self.context.loop_nonce_stack.pop();
-                        tracing::warn!("Exited loop.");
+                        tracing::trace!("Exited loop.");
                     }
                     Ordering::Equal => {}
                 }
 
-                tracing::warn!(
-                    loop_nonce = self.context.loop_nonce,
-                    stack_len = self.context.loop_nonce_stack.len()
-                );
-
                 self.context.subgraph_id = sg_id;
                 self.context.is_first_run_this_tick = sg_data
                     .last_tick_run_in
                     .is_none_or(|last_tick| last_tick < self.context.current_tick);
-                self.context.is_first_loop_iteration = self
-                    .context
-                    .loop_nonce_stack
-                    .last()
-                    .is_some_and(|&curr_nonce| sg_data.last_loop_nonce < curr_nonce);
 
+                if let Some(loop_id) = sg_data.loop_id {
+                    // If the previous execution of this subgraph had the same loop execution and
+                    // iteration count, then we need to increment the count.
+                    let curr_loop_nonce = self.context.loop_nonce_stack.last().copied();
+
+                    let curr_iter_count = if let Some(&(_loop_loop_nonce, loop_iter_count)) =
+                        self.loop_iter_counters.get(loop_id)
+                    {
+                        let (prev_loop_nonce, prev_iter_count) = sg_data.last_loop_nonce;
+
+                        // If the loop nonce is the same as the previous execution, then we are in
+                        // the same loop execution.
+                        // `curr_loop_nonce` is `None` for top-level loops, and top-level loops are
+                        // always in the same (singular) loop execution.
+                        if curr_loop_nonce.is_none_or(|nonce| nonce == prev_loop_nonce) {
```

Comment:
Under what situations will the same subgraph run multiple times in the same loop iteration? Since all inputs to a loop are bounded this shouldn't be needed?

Reply:
Correct, never more than once in a single iteration of the loop. But multiple times (once per iteration) in a loop "execution".

I'll make the comments more clear